### ISSUE TITLE
slow response during retrieve metrics

### ISSUE THREAD

**Issue by KVInventoR** [dmachard/dnstap-receiver#12]
I tried to get metrics from dnstap_receiver and got that time to get response is really long

my current statistics looks like

```
root@prometheus01:~# curl --user admin:changeme http://127.0.0.1:8080/metrics
# HELP dnstap_qps Number of queries per second received
# TYPE dnstap_qps counter
dnstap_qps 4841
# HELP dnstap_domains Number of domains asked
# TYPE dnstap_domains counter
dnstap_domains 5787
# HELP dnstap_clients Number of clients asked
# TYPE dnstap_clients counter
dnstap_clients 572
# HELP dnstap_queries Number of queries received
# TYPE dnstap_queries counter
dnstap_queries 53500
```

but time: time curl --user admin:changeme http://127.0.0.1:8080/metrics
real    0m24.688s
user    0m0.012s
sys     0m0.011s

24 seconds to retrieve statistics is really long time and it will also block node_exporter during retrieve metrics

**Comment by dmachard** [comment]
I don't succeed to reproduce in my side but my dnstap receiver does nothing. 

time curl --user admin:changeme http://10.0.0.51:8080/metrics
real    0m0.021s
user    0m0.003s
sys     0m0.001s

On your side, your dnstap receiver is under load ? if can you give me more details ?

**Comment by dmachard** [comment]
reproduced! I need more investigation in the code

**Comment by dmachard** [comment]
a complete redesign will be done for the next release with multithreading support

**Comment by KVInventoR** [comment]
sounds good
I like you project, it's really useful

**Comment by dmachard** [comment]
I made some improvements in the statistics engine, I no longer reproduce the slow reponse in my side. Could you retry from the source code ?

Also can you share some metrics like qps, cpu usage, memory usage and number of domains ? it will help me improve the performance of the dnstap receiver

**Comment by KVInventoR** [comment]
it's not high load dns server, it's our little installation, dnstap_receiver running directly on dns server.


dnstap_qps 2495
dnstap_domains 13201
dnstap_clients 14425
dnstap_queries 80901
dnstap_queries_udp 80618
dnstap_queries_tcp 283

```
top - 13:46:22 up 2 days, 17:48,  9 users,  load average: 0.68, 0.94, 0.95
Tasks: 156 total,   1 running, 155 sleeping,   0 stopped,   0 zombie
%Cpu(s):  4.8 us,  1.6 sy,  0.0 ni, 93.5 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem :   3946.4 total,    924.8 free,    511.6 used,   2510.0 buff/cache
MiB Swap:   2048.0 total,   2048.0 free,      0.0 used.   3130.9 avail Mem

```

**Comment by KVInventoR** [comment]
works really well

is there any reason to dublicate stats?

```
curl -s --user admin:changeme http://localhost:8080/metrics|grep "^dnstap_clients"
dnstap_clients 31812
dnstap_clients{identity="alpha-vm01"} 31823

```

probably it will be more useful to have dnstap_clients_total sum all client, 
the same for all another dublicated metrics
```
dnstap_qps 226
dnstap_qps{identity="alpha-vm01"} 226
```

**Comment by dmachard** [comment]
Good news :)

with my home lab, I can test only 5000 qps max so If you can tell how much the dnstap receiver is capable to receive, it will be great. I could add these informations in the README.md.

For the duplicate stats, the dnstap can handles multiple dnstap streams so in the metrics you have the total at the begenning and  after all stats for each dnstap streams.

I can renamed metrics to be more explicit, can you open a new issue for that ? 

Thanks for your feedbacks.

**Comment by dmachard** [comment]
fix in release 3.0.0
